File: interview-collections-algo/src/main/java/com/equalinformation/interview/algo/java/string_and_array/RotateArrayS1IntermediateArray_O_n.java
package com.equalinformation.interview.algo.java.string_and_array;

/**
 * Due to: ProgramCreek
 * Re-Created by bpupadhyaya on 3/5/16.
 *
 * Problem: Rotate an array of n elements to the right by k steps.
 * For example, with n = 5 and k = 2, the array [1,2,3,4,5] is
 * rotated to [4,5,1,2,3]. How many different ways do you know to
 * solve this problem?
 *
 * Solution note: Space is O(n) and time is O(n). You can check out the
 * difference between System.arraycopy() and Arrays.copyOf()
 */
public class RotateArrayS1IntermediateArray_O_n {

    public static void main(String...args) {
        RotateArrayS1IntermediateArray_O_n soln = new RotateArrayS1IntermediateArray_O_n();
        int[] givenArray = {1,2,3,4,5};
        int k = 2;
        soln.rotate(givenArray,k);
    }

    public void rotate(int[] nums, int k) {
        if(k > nums.length) {
            k = k % nums.length;
        }
        int[] result = new int[nums.length];

        for(int i = 0; i < k; i++) {
            result[i] = nums[nums.length - k + i];
        }

        int j = 0;
        for(int i = k; i < nums.length; i++) {
            result[i] = nums[j];
            j++;
        }

        System.arraycopy(result,0,nums,0,nums.length);

        System.out.println("Rotated array: ");
        for(int i: result) {
            System.out.print(i+", ");
        }
    }

}

File: interview-collections-algo/src/main/scala/com/equalinformation/interview/algo/scala/string_and_array/RotateArrayS1IntermediateArray_O_n.scala
package com.equalinformation.interview.algo.scala.string_and_array
import java.lang.System.arraycopy


/**
  * Created by bpupadhyaya on 6/12/16.
  */

object RotateArrayS1IntermediateArray_O_nTest {
  def main(args: Array[String]): Unit = {
    val soln = new RotateArrayS1IntermediateArray_O_n
    val givenArray = Array(1, 2, 3, 4, 5)
    val k = 2
    soln.rotate(givenArray, k)

  }
}

class RotateArrayS1IntermediateArray_O_n {

  def rotate(nums: Array[Int], k: Int): Unit = {
    var k_var = k;

    if(k_var > nums.length) {
      k_var = k_var % nums.length
    }

    val result = new Array[Int](nums.length)

    for(i <- 0 to k_var - 1 ) {
      result(i) = nums(nums.length - k_var + i)
    }

    var j: Int = 0
    for(i <- k_var to nums.length - 1 ) {
      result(i) = nums(j)
      j+= 1
    }

    arraycopy(result,0,nums,0,nums.length)

   System.out.println("Rotated array: ")
    for(i<- result) {
      print(i+", ")
    }

  }

}


